sys/aarch64/trap.c

```diff
@@ -47,52 +42,13 @@ static void syscall_handler(register_t code, ctx_t *ctx,
   result->error = error;
 }
 
-static void abort_handler(ctx_t *ctx, register_t esr, vaddr_t vaddr,
-                          bool usermode) {
-  uint32_t exception = ESR_ELx_EXCEPTION(esr);
-  thread_t *td = thread_self();
-
-  klog("%x at $%lx, caused by reference to $%lx!", exception, _REG(ctx, PC),
-       vaddr);
-
-  pmap_t *pmap = pmap_lookup(vaddr);
-  if (!pmap) {
-    klog("No physical map defined for %lx address!", vaddr);
-    goto fault;
-  }
-
+static vm_prot_t exc_access(u_long exc_code, register_t esr) {
   vm_prot_t access = VM_PROT_READ;
-  if (exception == EXCP_INSN_ABORT || exception == EXCP_INSN_ABORT_L) {
+  if (exc_code == EXCP_INSN_ABORT || exc_code == EXCP_INSN_ABORT_L)
     access |= VM_PROT_EXEC;
-  } else if (esr & ISS_DATA_WnR) {
+  else if (esr & ISS_DATA_WnR)
     access |= VM_PROT_WRITE;
-  }
-
-  int error = pmap_emulate_bits(pmap, vaddr, access);
-  if (error == 0)
-    return;
-
-  if (error == EACCES || error == EINVAL)
-    goto fault;
-
-  vm_map_t *vmap = vm_map_lookup(vaddr);
-  if (!vmap) {
-    klog("No virtual address space defined for %lx!", vaddr);
-    goto fault;
-  }
-  if (vm_page_fault(vmap, vaddr, access) == 0)
-    return;
-
-fault:
-  if (td->td_onfault) {
-    _REG(ctx, PC) = td->td_onfault;
-    td->td_onfault = 0;
-  } else if (usermode) {
-    /* Send a segmentation fault signal to the user program. */
-    sig_trap(ctx, SIGSEGV);
-  } else {
-    kernel_oops(ctx);
-  }
+  return access;
```

Comment:
I prefer brackets for more complex control flow like here.

Reply:
Done.